presto-pinot/src/main/java/io/prestosql/pinot/PinotBrokerPageSource.java

```diff
@@ -0,0 +1,433 @@
+        }
+    }
+
+    private void setValuesForGroupby(
+            List<BlockBuilder> blockBuilders,
+            List<Type> types,
+            Map<String, Integer> columnIndices,
+            Map<Integer, String> groupByFunctions,
+            List<String> groupByColumns,
+            JsonNode group,
+            String[] values)
+    {
+        checkState(groupByColumns.size() == group.size(), "Expected '%s' groupBy columns got '%s'", group.size(), groupByColumns.size());
+        for (int i = 0; i < group.size(); i++) {
+            int index = columnIndices.getOrDefault(groupByColumns.get(i), -1);
+            if (index >= 0) {
+                setValue(types.get(index), blockBuilders.get(index), group.get(i).asText());
+            }
+        }
+        for (int i = 0; i < values.length; i++) {
+            String columnName = groupByFunctions.get(i);
+            if (columnName != null) {
+                int index = columnIndices.getOrDefault(columnName, -1);
+                if (index >= 0) {
+                    setValue(types.get(index), blockBuilders.get(index), values[i]);
+                }
+            }
+        }
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return completedBytes;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return readTimeNanos;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finished;
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        if (finished) {
+            return null;
+        }
+
+        long start = System.nanoTime();
+        try {
+            List<Type> expectedTypes = columnHandles.stream()
+                    .map(PinotColumnHandle::getDataType)
+                    .collect(Collectors.toList());
+            PageBuilder pageBuilder = new PageBuilder(expectedTypes);
+            ImmutableList.Builder<BlockBuilder> columnBlockBuilders = ImmutableList.builder();
+            ImmutableList.Builder<Type> columnTypes = ImmutableList.builder();
+            ImmutableMap.Builder<String, Integer> columnIndices = ImmutableMap.builder();
+            for (int i = 0; i < columnHandles.size(); i++) {
+                BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(i);
+                columnBlockBuilders.add(blockBuilder);
+                columnTypes.add(expectedTypes.get(i));
+                columnIndices.put(columnHandles.get(i).getColumnName(), i);
+            }
+
+            int counter = issuePqlAndPopulate(
+                    brokerPql.getTable(),
+                    brokerPql.getPql(),
+                    brokerPql.getGroupByClauses(),
+                    columnBlockBuilders.build(),
+                    columnTypes.build(),
+                    columnIndices.build());
+            pageBuilder.declarePositions(counter);
+            Page page = pageBuilder.build();
+
+            // TODO: Implement chunking if the result set is ginormous
```

Comment:
Take a look at the Elasticsearch implementation for an example of how to paginate over the resultset to produce multiple Presto pages with adaptive sizing:

https://github.com/prestosql/presto/blob/master/presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/ElasticsearchPageSource.java#L166-L205

Reply:
The response is still retrieved in one chunk, but the page size is now adaptive. Will work with pinot folks to get the response in smaller chunks (like presto client). 

A limit can be enforced in PinotSplitManager::isBrokerQuery. Let me know if that makes sense.